### ISSUE TITLE
Activate keyless drive 

### ISSUE THREAD

**Issue by rjw2021** [Kemmey/Kemmey-TeslaWatch-Public#305]
I have Iphone 8 (latest OS), AppleWatch 4 Cellular (latest OS).  I have down loaded the Watch for Tesla.  I never carry my phone, so am so appreciative of this app.  All features on my AppleWatch work perfectly except for "Activate keyless drive".

1. I select the app.  Tesla wakes up.
2. I select the ICON/Remote Start and receive "Token v3 received".
3. But my new 2021 Model Y does not recognize the key. I press the brake and get the warning "no key in the vehicle".

Any suggestions?

I am very curious technically how all this works.  My AppleWatch via Cellular, my apple watch connects to....then....(How does the token get to the vehicle?.

Thanks. Richard

**Comment by Kemmey** [comment]
Hey Richard,

This sounds to me like the password entered in settings under Remote Start has a typo. This has to be your Tesla account password - it's easier to enter in the watch app for Tesla on your phone and let it sync automatically to your watch. Please try to type it in again, wait a few minutes with your phone and watch nearby and then try remote start again. 

As for how the tech works - it's pretty simple - when you initially installed the apps and logged in, you generated an access token to the Tesla servers. That token has a couple parts, one is the actual key for getting access - that's your access token. It's also got a refresh token - that's good for getting a new access token, once that expires. It expires approx every 45 days. With the refresh token, a fresh access token can be issued whenever needed. Sometimes the watch will also just grab the token from your phone - this happens s if the token is missing from the watch. 

What you experience, is actually the watch logging you out because of your remote start password being wrong, which actually throws you out of the app - however, at the login screen, the watch just grabs your login from your phone, and logs right back in. That's actually bug! You should remain logged out... 

Long story short - triple check your Tesla account password is entered in the remote start section in settings :-D 

Cheers,

Kim

**Comment by rjw2021** [comment]
Kim,

Still think it is so cool to talk directly to the developer.

Your app is AWESOME and works perfectly!!  Thanks for the effort.

I do have an issue of it draining my AppleWatch battery.  I have settings set so it doesn’t always connect and wake up the car.

If I don’t use it at all for the day, I am good to go.

However, if I even activate it from my watch one time, it seems to suck my AppleWatch power and my watch doesn’t make it through the day.

Any suggestions?  

P.S. I don’t like carrying my phone.  I always leave it in the car, so your app is AWESOME for someone like me.

Thanks.

Richard

------------------------------------------

**Comment by Kemmey** [comment]
Hey Richard, 

I'm a one-man-shop so all you got is me! :-D Happy to help with anything!

I must admit that I've noticed some battery drain recently myself, so I'm currently trying to optimize a few things to ensure the app is more battery efficient. 

Unfortunately, watchOS does not capture app power usage statistics - or at least, it doesn't expose that data like iOS does. It would be extremely valuable information to me, but it's unfortunately inaccessible. 

If just launching the app a single time drains your watch battery for the day, something is definitely going wrong. In previous versions, I let the backgrounding execute for up to a minute, but the power watchdog started killing the process for excessive use - so you're definitely on to something - but in recent releases I've reduced background processing time to under 10 seconds. After that, it should give up by itself. 

Can you verify if recent releases have improved - or is this still the case, that a single launch kills your battery? 

Can you tell me what watch you have - model, size and age? Also, if you go to the Settings app, tap Battery, Battery Health, you should be able to see you watch'es maximum capacity - my close-to-a-year-old series 6 44mm is already degraded to 93%! The batteries in these things definitely don't age well... 

Cheers, 

Kim
